modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/GsubProcessor.java

```diff
@@ -49,7 +49,17 @@ String getReplacement() {
 
     @Override
     protected String process(String value) {
-        return pattern.matcher(value).replaceAll(replacement);
+        try {
+            return pattern.matcher(value).replaceAll(replacement);
+        } catch (StackOverflowError e) {
+            /*
+             * A bad regex on problematic data can trigger a StackOverflowError. In this case we can safely recover from the
+             * StackOverflowError, so we rethrow it as an Exception instead. This way the document fails this processor, but processing
+             * can carry on. The value would be useful to log here, but we do not do so for because we do not want to write potentially
+             * sensitive data to the logs.
+             */
+            throw logAndBuildException("Caught a StackOverflowError while processing gsub pattern: [" + pattern + "]", e);
```

Comment:
These errors are exceedingly rare in practice (like I've seen evidence of something like this maybe 4 times in 3 years). It's a combination of a bad regex + bad data. We can't log the data, but logging the regex at least gives us and our support team _something_ to go by. And it could very well be Elastic's own regexes from some integration causing the failure. Right now we just have nothing to go by. We could potentially throttle this, or write these failures to a custom index or something, but that all seems like overkill for something that's pretty rare, doesn't it?

Reply:
Before this PR, the StackOverflowError takes down the server and the user received no response. That's why I'm pretty sure it's rare. We can potentially change logAndBuildException to log a different message -- or are you saying that we ought to not be logging anything on this kind of error? It's arguably a bug in the system that we allow a StackOverflowError to happen in the first place inside of the database process, so it seems log-worthy. For example if we start seeing this a lot in the logs we might want to prioritize a different fix.
Just to tie them together, the details of how logAndBuildException works were worked out as part of https://github.com/elastic/elasticsearch/pull/99493.

Reply:
I'm not against catching the SOE, we should definitely do that. But I don't think logging the bad regex is necessary. It's a user concern: it's their bad regex. The fact that we must resort to catching SOE is an unfortunate result of using java's Pattern. In painless we try to be more aggressive in rejecting complex regexes (see https://github.com/elastic/elasticsearch/pull/63029), but even then we still catch SOE there.

Reply:
> but I don't think logging the bad regex is necessary. It's a user concern: it's their bad regex.

If I only have an errored document, or a log that tells me "Found bad regex taking too much time, would have stackoverflowed" it's really hard to debug. I had to look through ~200 GROK patterns and ~100 gsubs to find the offending pattern. Logging the bad regex should be very helpful because with that I can search my pipelines and I know which one is broken.

Reply:
The bad regex would still be returned in the error though? I'm not against making the failure contain the bad regex, only logging it at the server. We don't log other "user problems" (ie if a user sends something resulting in a 4xx, we don't log that fact).

Reply:
I wouldn't be shocked to find one of own integrations doing this. In that case we'd have to rely on users noticing and giving us their error messages. Is your worry that these regexes are going to be sensitive since they could be user-generated, and shouldn't be in logs?

Reply:
I share Ryan's concerns here. We should push errors like this back onto the client, which (a) knows a lot more context about the erroneous request than we do within ES and (b) can actually do something to address the problem. If it's our own integrations then we can rely on them handling or exposing this sort of error properly, including sharing all the context needed to work out where things are going wrong. That would help the support team engage directly with the folks that own the problem rather than unnecessarily escalating things to the ES team.

If nobody is noticing these errors then the situation seems fairly hopeless no matter where these things are logged.

Reply:
Maybe a silly question:
* Elastic Agent / Logstash sends a doc in a bulk request to an ingest pipeline. Ingest pipeline contains a bad regex. How would that look like? It will be a failed request in the bulk and therefore agent / logstash should log it, similiarly to a mapping exception?

@ruflin do you know how this here will impact the logs+ with the failure store inside ES instead of logging back to the agent that there was an issue with this document. Would the document then contain a `error.ingest_failure` or something like that telling us what went wrong?

Reply:
In general I will expect when the failure store is enabled, the "bad" document will end up in the failure store and ideally the document itself will contain some info around what broke. What is returned to the client, I need @dakrone to chime on this one.

Reply:
I think this sounds like a good use for the failure store (and perhaps overlapping with those users that would be likely to use the failure store in the first place). Yes, the failure store document will contain the exception if thrown during indexing (assuming it's enabled of course). I could see a situation where we log it, and a user sends many of these and they pollute the logs a lot. Logging anything on a per-document basis can potentially be really large.

What about changing the logging to be trace-only (so it's still accessible if absolutely necessary), and still returning the message in the regular thrown exception (so it makes it into the failure store)?

Reply:
Alright, I changed to trace-only logging, with the regex in the exception message and in the trace-only log message. Does that sound ok to everyone else?